docs/WASI-design-principles.md

```diff
@@ -0,0 +1,95 @@
+
+Note that this is a different sense of "capability" than [Linux capabilities](http://man7.org/linux/man-pages/man7/capabilities.7.html) or the withdrawn [POSIX capabilities](https://archive.org/details/posix_1003.1e-990310), which are per-process rather than per-resource.
+
+The simplest representation of handles are values of [reference type](https://github.com/WebAssembly/reference-types). References in wasm are inherently unforgeable, so they can represent handles directly.
+
+Some programming languages operate primarily within linear memory, such as C, C++, and Rust, and there currently is no easy way for these languages to use references in normal code. And even if it does become possible, it's likely that source code will still require annotations to fully opt into references, so it won't always be feasible to use. For these languages, references are stored in a [table](https://webassembly.github.io/spec/core/bikeshed/index.html#tables%E2%91%A0) called a *c-list*. Integer indices into the table then identify resources, which can be easily passed around or stored in memory. In some contexts, these indices are called *file descriptors* since they're similar to what POSIX uses that term for. There are even some tools, such as wasm-bindgen, which make this fairly easy. (Internally, tools and engines don't always use actual WebAssembly tables to do this, however those are implementation details. Conceptually, they work as if they had tables.)
+
+If all instances shared a single c-list, that would undermine isolation, because any instance could forge an index to any other instance's resources. Fortunately, there can be multiple c-lists, and instances can be given access to some c-lists and not others, or even have instances which don't use any c-lists at all, so it's still possible to establish isolation between instances.
+
+Witx-specified APIs use a special `handle` keyword to mark parameters and return values which are handles. In the short term, these are lowered to integer indices, with an implied table, so that the APIs can be easily used from C and similar languages today. Once [interface types](https://github.com/WebAssembly/interface-types) and [type imports](https://github.com/WebAssembly/proposal-type-imports) are ready, we expect to make use of them to provide APIs which can be used either from languages using references or from languages using integer indices, with tables being used and managed automatically.
+
```

Comment:
Thoughts on this proposal https://github.com/WebAssembly/WASI/issues/109#issuecomment-542921814 - would that fit in this model?

Reply:
Specifically, I think it would be good to address how the standard POSIX etc. model of "get me a handle for this string path" would work - who would be responsible for that, would it be in the VM or in wasm or in a polyfill (as in that proposal) or somewhere else.

Reply:
I added some text [here](https://github.com/WebAssembly/WASI/pull/192/commits/baaa84fa878ad86eac2fa661662dd3716a7e8a30) which gives an overview of how this would look. Does that answer your question?

Reply:
Not fully: it has words like "de-emphasizing" and "typically", so I'm not sure how to interpret it. Concretely, again, I'm asking if this document would accept @tlively's compromise proposal, or not (or "unsure").

Reply:
Specifically, I think the ambiguity comes from the sentence `When needed, compatibility with POSIX-style APIs is then provided as a thin layer on top implementing a simple name-to-handle mapping.` This sentence does not indicate which party should be responsible for providing this layer. There would be a benefit to code size and portability if WASI environments had to provide these compatibility layers as standardized imports in much the same way that normal WASI functions are standardized imports. The difference would be that the compatibility layers would be implemented in user space rather than by the engine to preserve OCAP.

Reply:
Are you talking about [this](https://github.com/WebAssembly/WASI/issues/109#issuecomment-542921814)? I'll respond in that issue.

Reply:
Yes, that's what we're thinking of. Follow-up discussion there sounds good.